In [1]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import scipy.stats

%matplotlib inline

In [2]:
matplotlib.rc('font', family = 'Malgun Gothic')
plt.rc('font', family = 'Malgun Gothic')

## Weibull분포

In [34]:
x = 2500
alpha = 5.2
beta = 2200

#누적확률 값 계산 
prob = stats.weibull_min.cdf(x,alpha,scale = beta)
print("P(X>={0}):{1:.3f}".format(x,(1-prob)))

P(X>=2500):0.143


전체 넓이 - f(X<=2500) = 1 - prob  

※ 결론: 제품이 2500시간 이상 작동할 확률은 0.143이다.

## 이항분포

In [4]:
n = 5

for i in range(n+1):
    
    prob =stats.binom.pmf(k=i,n=n,p=0.3)
    print("P(X={0})={1:.3f}".format(i,prob))

P(X=0)=0.168
P(X=1)=0.360
P(X=2)=0.309
P(X=3)=0.132
P(X=4)=0.028
P(X=5)=0.002


※ 결론 : 5대의 차량을 임의로 선택했을때 불량대수가 0,1,2,3,4,5 대가 나올 확률은 위와 같다. 

## 포아송분포

In [5]:
m = 3
prob = stats.poisson.pmf(2,m)


print("{0}명의 손님이 도착할 확률 : {1:.1f}".format(2,prob*100))

2명의 손님이 도착할 확률 : 22.4


In [9]:
m = 3
prob = stats.poisson.pmf(1,m)


print("{0}명 이하의 손님이 도착할 확률 : {1:.1f}".format(1,(prob)*100))

1명 이하의 손님이 도착할 확률 : 14.9


In [8]:
m = 3
prob = stats.poisson.pmf(3,m)


print("{0}명 이하의 손님이 도착할 확률 : {1:.1f}".format(3,(1-prob)*100))

3명 이하의 손님이 도착할 확률 : 77.6


※결론 : 일정시간(10시에서 11시)사이에 각 조건의 손님이 도착할 확률은 위와 같다. 

## 정규성검정

In [80]:
ds_mat = pd.read_csv("data/2sample_t_test.csv",engine = "python")
ds_mat.head()

Mat_A  Mat_B  Total  Sub
0   73.4   68.7   73.4    1
1   77.0   71.4   77.0    1
2   73.7   69.8   73.7    1
3   73.3   75.3   73.3    1
4   73.1   71.3   73.1    1

In [86]:
data1 = ds_mat['Mat_A'][0:10]
data1 

0    73.4
1    77.0
2    73.7
3    73.3
4    73.1
5    71.5
6    74.5
7    77.5
8    76.4
9    77.7
Name: Mat_A, dtype: float64

In [87]:
statistic,p = stats.shapiro(data1)
print("Shapiro = Wilk Test : statistic={},p-value={}".format(statistic,p))

Shapiro = Wilk Test : statistic=0.9062636494636536,p-value=0.2562994360923767


귀무가설 :정규성을 만족한다.   
대립가설 :정규성을 만족하지 않는다.  

※결론 : p값이 0.256으로 유의수준(alpha=0.05)보다 크므로 귀무가설을 채택한다.(=data1은 정규성을 만족한다.)

In [83]:
data2 = ds_mat['Mat_B'][0:10]
data2

0    68.7
1    71.4
2    69.8
3    75.3
4    71.3
5    72.7
6    66.9
7    70.2
8    74.4
9    70.1
Name: Mat_B, dtype: float64

In [79]:
statistic,p = stats.shapiro(data2)
print("Shapiro = Wilk Test : statistic={},p-value={}".format(statistic,p))

Shapiro = Wilk Test : statistic=0.9729753732681274,p-value=0.9169777035713196


귀무가설 : 정규성을 만족한다.  
대립가설 : 정규성을 만족하지 않는다.   

※결론 : p값이 0.917으로 유의수준(alpha=0.05)보다 크므로 귀무가설을 채택한다.(=data2는 정규성을 만족한다.)

## 모평균추정(모표준편차 아는 경우)

In [15]:
df = pd.DataFrame({"sample":[18,18,20,21,20,23,19,18,17,21,22,20,20,21,20,19,19,18,17,19]})
df.head()

sample
0      18
1      18
2      20
3      21
4      20

In [16]:
lower,upper = stats.norm.interval(0.95, loc = np.mean(df),scale = 3.8/np.sqrt(20))
print("신뢰구간:{0},{1}".format(lower.round(2),upper.round(2)))

신뢰구간:[17.83],[21.17]


※결론 : 평균무게에 대한 95% 신뢰구간은 17.83 ~ 21.17 이다. 

## 모평균추정(모표준편차 모르는 경우)

In [17]:
df =pd.DataFrame({'sample':[73,71,74,69,70,73,70,68,75,72,70,72,73,70,70,72,71,70,75,72]})
lower,upper = stats.t.interval(0.95,len(df)-1, loc =np.mean(df),scale =scipy.stats.sem(df))
print("신뢰구간:{0},{1}".format(lower.round(2),upper.round(2)))

신뢰구간:[70.6],[72.4]


※결론 : 모평균의 95% 신뢰구간은 70.6 ~ 72.4 이다.

## 모분산 신뢰구간 추정

In [45]:
df = pd.DataFrame({"sample":[12,9,10,5,16,13,5,9,8,13,16,12,12,4,14,9,11,14,9,17,5,8,9,6,11,19,6,6,8,9,13,10,8,13,19,8,12,14,16,16,11,17,5,9,12,17,11,8,5,8]})
df.describe()

sample
count  50.000000
mean   10.740000
std     3.983691
min     4.000000
25%     8.000000
50%    10.500000
75%    13.000000
max    19.000000

=> std의 값은 3.98

In [92]:
n = 50
std = 3.98

# chai1  =>  x^2(0.05/2,49) = x^2(0.025,49) =>카이분포표(alpha = 0.025,df =49)에서 71.42
# chai2  =>  x^2((1 - 0.05/2),49) = x^2(0.975,49) =>카이분포표(alpha = 0.975,df =49에서 32.36

chai1 = 71.42
chai2 = 32.36

a = (n - 1) * std**2 / chai1
b = (n - 1) * std**2 / chai2

print("모분산의 95% 신뢰 구간 : ({0:.2f},{1:.2f})".format(a, b)) 

모분산의 95% 신뢰 구간 : (10.87,23.99)


## 모비율의 신뢰구간 추정

In [90]:
n = 50 
phat = 1 - (9 / 50)

# z(0.025) => 정규분포표에서 1.96  
z = 1.96 

lower = phat - z * np.sqrt(phat * (1 - phat) / n)
upper = phat + z * np.sqrt(phat * (1 - phat) / n)

print("통화 관리 성공율에 대한 95% 신뢰 구간 :({0}, {1})".format(lower.round(2), upper.round(2)))


통화 관리 성공율에 대한 95% 신뢰 구간 :(0.71, 0.93)


## 평균검정(모집단 표준편차를 모르는 경우)

In [94]:
df = pd.DataFrame({'sample':[85.0,79.0,79.1,79.9,81.6,78.6,85.4,83.4,78.1,79.2]})
t_result = stats.ttest_1samp(df,78.0)

t,p = t_result.statistic.round(3), t_result.pvalue.round(3)

print("1-sample t-test")
print("t통계량 :{}".format(t))
print("p-value:{}".format(p))

1-sample t-test
t통계량 :[3.379]
p-value:[0.008]


귀무가설 : 개선활동이 만족도를 변화시켰다.  
대립가설 : 개선활동이 만족도를 변화시키지 않는다.  

※ 결론 : p -value 값이 0.008으로 유의수준(alpha = 0.05) 보다 작으므로 귀무가설을 기각한다. (=개선활동이 만족도를 변화시키지 않는다.)

## 2 Sample t-test

In [66]:
df1 = pd.DataFrame({'sample':[1.883,1.715,1.799,1.768,1.711,1.832,1.427,1.344]})
df2 = pd.DataFrame({'sample':[1.435,1.572,1.486,1.511,1.457,1.548,1.404,1.883]})

t_result =stats.ttest_ind(df1,df2)

t,p = t_result.statistic.round(3),t_result.pvalue.round(3)

print("2-sample t-test")
print("t:{}".format(t))
print("p:{}".format(p))

2-sample t-test
t:[1.7]
p:[0.111]


귀무가설 : 공급자 간에 모평균의 차이가 있다.  
대립가설 : 공급자 간에 모평균의 차이가 없다.   


※결론 : 유의수준 5% 에서 검정결과 P값이 0.111 으로 alpha(=0.05)보다 크므로 귀무가설을 기각한다.(=공급자간에 모평균에 차이가 없다.)

## paired t-test

In [71]:
df1 = pd.DataFrame({'sample': [720, 589, 780, 648, 720, 589, 780, 648, 780, 648]})
df2 = pd.DataFrame({'sample': [710, 580, 787, 712, 750, 600, 782, 670, 790, 680]})

t_result = stats.ttest_rel(df1,df2)

t,p = t_result.statistic.round(3), t_result.pvalue.round(3)

print(" Paired t-test")
print(" t 검정통계량 = %.3f \n p-value = %.3f" % t_result)

 Paired t-test
 t 검정통계량 = -2.266 
 p-value = 0.050


귀무가설 : 학원가기 전후에 대해 차이가 없다.   
대립가설 : 학원가기 전후에 대해 차이가 있다.  

※ 결론 : 유의수준 5%에서 검정결과 P값이 0.05 으로 (alpha=0.05)보다 작지 않으므로 귀무가설을 채택한다.(=학원가기 전후에 대해 차이가 없다.)

## 1 Proportion test

In [29]:
from statsmodels.stats.proportion import proportions_ztest

In [30]:
count = 40
n_obs = 100
value = 0.5

stat,pval = proportions_ztest(count,n_obs,value)

print("1 Proportion test")
print("z :{0:0.3f}".format(stat))
print('p:{0:0.3f}'.format(pval))

1 Proportion test
z :-2.041
p:0.041


귀무가설 : p(동전의 앞면이 나올 확률) = 0.5 이다.(해당 동전이 정상적이다.)  
대립가설 : p = 0.5가 아니다. (해당 동전은 정상적이지 않다.)  

※결론 : P값이 0.041 으로 (alpha = 0.05)보다 작으므로 귀무가설을 기각한다.(=해당동전은 정상 동전이 아니다.) 

## 2 Proportion test

In [31]:
count = np.array([14,5])
nobs = np.array([1200,1200])

stat, pval = proportions_ztest(count,nobs)

print("2 Proportion test")
print("z :{0:0.3f}".format(stat))
print('p:{0:0.3f}'.format(pval))

2 Proportion test
z :2.073
p:0.038


귀무가설 : 두 공정의 불량률이 같다.   
대립가설 : 두 공정의 불량률이 같지 않다.     

※결론 : p값이 0.038으로 (alpha=0.05)보다 작으므로 귀무가설을 기각한다.(= 두 공정의 불량룰은 같지 않다.) 